### DISCUSSION TITLE
Lost event occur when increase the BPFTRACE_MAX_PROBES

### DISCUSSION THREAD

**Discussion by miloluo** [bpftrace/bpftrace#2111] [Q&A]
Hi, Expert
This problem comes from  [BPFTRACE_MAX_PROBES increase and coredump occurred #2104 ](https://github.com/iovisor/bpftrace/discussions/2104)

When using the solution for fix the coredump issue while increase the  BPFTRACE_MAX_PROBES  environment variable.

```
 [root@dsc19c-n1 ~]# export BPFTRACE_MAX_PROBES=620
[root@dsc19c-n1 ~]# ulimit -n 4096
[root@dsc19c-n1 ~]#  bpftrace -e 'tracepoint:syscalls:* { printf("%s [%d: %s]  \n", strftime("%H:%M:%S", nsecs), pid,comm); }'
Attaching 612 probes...
10:31:41 [10541: sh]  
10:31:41 [10541: sh]  
10:31:41 [10541: sh]  
10:31:41 [10541: sh]  
10:31:41 [10541: sh]  
10:31:41 [10541: rm]  
10:31:41 [10531: bpftrace]  
10:31:41 [10531: bpftrace]  
10:31:41 [10531: bpftrace]  
10:31:41 [10531: bpftrace]  
10:31:41 [10531: bpftrace]  
10:31:41 [10531: bpftrace]  
10:31:41 [10531: bpftrace]  
10:31:41 [10531: bpftrace]  
......
10:31:41 [9186: ios_vktm_+ios1]  
10:31:41 [8791: apx_vktm_+apx1]  
Lost 59177 events
10:31:43 [10531: bpftrace]  
Lost 1 events
10:31:43 [10531: bpftrace]  
Lost 3 events
10:31:43 [10531: bpftrace]  
Lost 1 events
10:31:43 [10531: bpftrace]  
.....
10:31:44 [10531: bpftrace]  
Lost 3 events
10:31:44 [10531: bpftrace]  
Lost 3 events
10:31:44 [10531: bpftrace]  
Lost 3 events
10:31:44 [10531: bpftrace]  
Lost 3 events
10:31:44 [10531: bpftrace]  
Lost 3 events
10:31:44 [10531: bpftrace]  
Lost 3 events
^C10:31:44 [10531: bpftrace]   <<<<<<<<<< I issue ctrl + c here, so from this point the output is not very neat.
Lost 3 events
Lost 3 events
Lost 3 events
Lost 3 events
Lost 3 events
Lost 3 events
Lost 3 events
Lost 3 events
Lost 3 events
Lost 3 events
Lost 3 events
Lost 3 events
Lost 3 events
Lost 3 events
Lost 3 events
Lost 3 events
Lost 3 events
Lost 3 events
Lost 3 events
Lost 3 events
Lost 3 events
```

Could you help to how to solve this problem?

Thanks very much.

**Comment by fbs** [comment]
this happens when the kernel side cannot write to the ring buffer any more due to it being full. Right now bpftrace is single threaded and just processes one event at a time. Tracing all system calls on a busy multi core machine can easily overload bpftrace and lead to dropped events.

If i look at my basically idle repo server, doing 0 requests per second, only my ssh session and a docker daemon without containers:

```
# ulimit -n 1700; BPFTRACE_MAX_PROBES=620 bpftrace -e 'tracepoint:syscalls:* { @count=count() } i:s:1 { print(@count); clear(@count); }'
Attaching 613 probes...
@count: 10148
@count: 236
@count: 232
@count: 300
@count: 266
@count: 240
@count: 232
@count: 238
@count: 238
@count: 254
@count: 240
```

So thats already 250 syscalls on a system that basically does nothing.

On the bpftrace side there is always room for improvement, and patches are welcome :). But the best approach is to reduce the events, try filtering by pid, aggregating in maps or anything else you can think of.

One thing you can try is increasing `BPFTRACE_PERF_RB_PAGES` but that just ups the buffer for bursts

**Reply by miloluo** [reply]
Hi, Expert
I understand what you mean, but actually, from the start point, we should find which processes and events we should focus, and then we can do this to filter job. I mean first we should find which processes we should focus and which events we should focus(may be it's too slow or it's related  )
Do you have this suggestion on this?

Thansk very much.
